packages/adblocker/src/engine/bucket/preprocessor.ts

```diff
@@ -0,0 +1,136 @@
+import { StaticDataView } from '../../data-view';
+import IFilter from '../../filters/interface';
+import Preprocessor, { Env } from '../../preprocessor';
+
+type FilterId = number;
+
+export default class PreprocessorBucket {
+  public static deserialize(view: StaticDataView): PreprocessorBucket {
+    const excluded = new Set<FilterId>();
+    for (let i = 0, l = view.getUint32(); i < l; i++) {
+      excluded.add(view.getUint32());
+    }
+
+    const preprocessors: Preprocessor[] = [];
+    for (let i = 0, l = view.getUint32(); i < l; i++) {
+      preprocessors.push(Preprocessor.deserialize(view));
+    }
+
+    return new this({
+      excluded,
+      preprocessors,
+    });
+  }
+
+  private readonly preprocessors: Preprocessor[];
+  private readonly excluded: Set<FilterId>;
+
+  constructor({
+    excluded = new Set(),
+    preprocessors = [],
+  }: {
+    excluded?: Set<FilterId>;
+    preprocessors?: Preprocessor[];
+  }) {
+    this.excluded = excluded;
+    this.preprocessors = preprocessors;
+  }
+
+  public isFilterExcluded(filter: IFilter) {
+    return this.excluded.has(filter.getId());
+  }
+
+  public updateEnv(env: Env) {
+    // Update excluded filter ids based on bindings
+    this.excluded.clear();
+
+    for (let i = 0; i < this.preprocessors.length; ) {
```

Comment:
Why do we still need this `i` variable? Could we do with something simpler? I think it'd be best to not add empty preprocessors by construction instead of doing the check here on `updateEnv`.
```suggestion
    for (const preprocessor of this.preprocessors) {
```